File: connection/TypeDBSessionImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {Database} from "../api/connection/database/Database";
import {TypeDBOptions} from "../api/connection/TypeDBOptions";
import {SessionType, TypeDBSession} from "../api/connection/TypeDBSession";
import {TransactionType, TypeDBTransaction} from "../api/connection/TypeDBTransaction";
import {ErrorMessage} from "../common/errors/ErrorMessage";
import {TypeDBClientError} from "../common/errors/TypeDBClientError";
import {RequestBuilder} from "../common/rpc/RequestBuilder";
import {TypeDBStub} from "../common/rpc/TypeDBStub";
import {RequestTransmitter} from "../stream/RequestTransmitter";
import {TypeDBTransactionImpl} from "./TypeDBTransactionImpl";
import {TypeDBClientImpl} from "./TypeDBClientImpl";
import SESSION_CLOSED = ErrorMessage.Client.SESSION_CLOSED;

export class TypeDBSessionImpl implements TypeDBSession {

    private readonly _databaseName: string;
    private readonly _type: SessionType;
    private readonly _options: TypeDBOptions;
    private readonly _client: TypeDBClientImpl;
    private _id: string;
    private _database: Database;
    private _isOpen: boolean;
    private _pulse: NodeJS.Timeout;
    private _networkLatencyMillis: number;
    private _transactions: Set<TypeDBTransaction.Extended>;

    constructor(database: string, type: SessionType, options: TypeDBOptions, client: TypeDBClientImpl) {
        this._databaseName = database;
        this._type = type;
        this._options = options;
        this._client = client;
        this._isOpen = false;
        this._transactions = new Set();
    }

    async open(): Promise<void> {
        const openReq = RequestBuilder.Session.openReq(this._databaseName, this._type.proto(), this._options.proto())
        this._database = await this._client.databases.get(this._databaseName);
        const start = (new Date()).getMilliseconds();
        const res = await this._client.stub().sessionOpen(openReq);
        const end = (new Date()).getMilliseconds();
        this._id = res.getSessionId_asB64();
        this._networkLatencyMillis = Math.max((end - start) - res.getServerDurationMillis(), 1);
        this._isOpen = true;
        this._pulse = setTimeout(() => this.pulse(), 5000);
    }

    async close(): Promise<void> {
        if (this._isOpen) {
            this._isOpen = false;
            for (const tx of this._transactions) {
                await tx.close();
            }
            this._client.closeSession(this);
            clearTimeout(this._pulse);
            const req = RequestBuilder.Session.closeReq(this._id);
            await this._client.stub().sessionClose(req);
        }
    }

    async transaction(type: TransactionType, options?: TypeDBOptions): Promise<TypeDBTransaction> {
        if (!this.isOpen()) throw new TypeDBClientError(SESSION_CLOSED);
        if (!options) options = TypeDBOptions.core();
        const transaction = new TypeDBTransactionImpl(this, type, options);
        await transaction.open();
        this._transactions.add(transaction);
        return transaction;
    }

    get database(): Database {
        return this._database;
    }

    isOpen(): boolean {
        return this._isOpen;
    }

    get options(): TypeDBOptions {
        return this._options;
    }

    get type(): SessionType {
        return this._type;
    }

    get id() {
        return this._id;
    }

    get stub(): TypeDBStub {
        return this._client.stub();
    }

    get requestTransmitter(): RequestTransmitter {
        return this._client.transmitter();
    }

    get networkLatency() {
        return this._networkLatencyMillis;
    }

    private async pulse(): Promise<void> {
        if (!this._isOpen) return;
        const pulse = RequestBuilder.Session.pulseReq(this._id);
        try {
            const isAlive = await this._client.stub().sessionPulse(pulse);
            if (!isAlive) this._isOpen = false;
            else this._pulse = setTimeout(() => this.pulse(), 5000);
        } catch (e) {
            this._isOpen = false;
        }
    }
}

File: connection/TypeDBSessionImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.connection;

import com.vaticle.typedb.client.api.TypeDBOptions;
import com.vaticle.typedb.client.api.TypeDBSession;
import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.database.Database;

import java.util.concurrent.atomic.AtomicBoolean;

import static com.vaticle.typedb.client.jni.typedb_client.session_get_database_name;
import static com.vaticle.typedb.client.jni.typedb_client.session_new;
import static com.vaticle.typedb.client.jni.typedb_client.session_on_close;

public class TypeDBSessionImpl implements TypeDBSession {
    public com.vaticle.typedb.client.jni.Session session;
    private final Type type;
    private final TypeDBOptions options;
    private final AtomicBoolean isOpen;

    TypeDBSessionImpl(Database database, Type type, TypeDBOptions options) {
        this.session = session_new(((TypeDBDatabaseImpl) database).database.released(), type.asJNI(), options.options);
        this.type = type;
        this.options = options;
        isOpen = new AtomicBoolean(true);
    }

    @Override
    public boolean isOpen() {
        return isOpen.get();
    }

    @Override
    public Type type() {
        return type;
    }

    @Override
    public String database_name() {
        return session_get_database_name(session);
    }

    @Override
    public TypeDBOptions options() {
        return options;
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type) {
        return transaction(type, new TypeDBOptions());
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type, TypeDBOptions options) {
        return new TypeDBTransactionImpl(this, type, options);
    }

    @Override
    public void onClose(Runnable function) {
        session_on_close(session, new Callback(function).released());
    }

    @Override
    public void close() {
        if (isOpen.compareAndSet(true, false)) {
            session.delete();
            session = null;
        }
    }

    static class Callback extends com.vaticle.typedb.client.jni.SessionCallbackDirector {
        private final Runnable function;

        Callback(Runnable function) {
            this.function = function;
        }

        @Override
        public void callback() {
            function.run();
        }
    }
}
